common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/BodyProcessorCreator.java

```diff
@@ -77,7 +83,10 @@ public void setValue(RestClientRequest clientRequest, Object arg) throws Excepti
       try (BufferOutputStream output = new BufferOutputStream()) {
         clientRequest.putHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
         RestObjectMapper.INSTANCE.writeValue(output, arg);
-        clientRequest.write(output.getBuffer());
+        if (arg != null) {
```

Comment:
Can you add a test case to show return body is required while users pass a null value? Json will encoding null to "null", so I think here may got some problem.

Reply:
its already added. private static void testRequiredBody(RestTemplate template, String microserviceName)   , server code @ AnnotationsTest.java,
TestMgr.check("No user data found",
+        template.postForObject(prefix + "/annotations/saysomething?prefix={prefix}",
+            null,

Reply:
I think there are still something missed. Let us view the consumer as a browser or postman.

Assume we have a service:
   void post(Person user)
Before this feature:
post man body: 
1. empty: provider will report invalidate parameter
2. null: (string null): provider will get a null Person instance

After this change:
1. empty: provider will get a null Person
2. null: provider will get a null Person(I am not tested)

And here you add if(arg !=null) statement, 
Before:
consumer send body: null string

Now: consumer send: empty 

This is a change, maybe you need to a manual test using postman. 

For null objects, consumer send empty is better than null, I think, so I can accept this change. But there are one consequence, that when required=true, users can send null string to provider but after change, user can never pass null to Person. 

Can you test and confirm? If my guess is correct, this code can not change, that's consumer always send null to provider for objects. required only work for 3rdparty clients.

Reply:
Maybe can pass "" as empty body using RestTemplate, but it's a bit strange send "" can not cast to Person.

Reply:
when required = true
1. Resttempalte or RPC always send empty body to provider and provider throws Exception
2. postman can send empty or null to provider, and provider throws Exception
when required = false
1. Resttempalte or RPC always send empty body to provider and provider get null
2. postman can send empty or null to provider, and provider get null

After change this is the expected value. I think is acceptable. When users not specify required = true, we can default to false.